toolkit/tools/imagegen/diskutils/initramfs.go

```diff
@@ -0,0 +1,229 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// Utility to encrypt disks and partitions
+
+package diskutils
+
+import (
+	"bytes"
+	"io"
+	"os"
+
+	"github.com/cavaliercoder/go-cpio"
+	"github.com/klauspost/pgzip"
+	"microsoft.com/pkggen/internal/logger"
+)
+
+// InitramfsMount represented an editable initramfs
+type InitramfsMount struct {
+	pgzWriter           *pgzip.Writer
+	cpioWriter          *cpio.Writer
+	outputBuffer        *bytes.Buffer
+	initramfsOutputFile *os.File
+}
+
+// CreateInitramfs creates a new initramfs
+func CreateInitramfs(initramfsPath string) (initramfs InitramfsMount, err error) {
+	// Initramfs traditionally is -rw-------
+	const initramfsModeBits = os.FileMode(0600)
+	// Caller must use InitramfsMount.Close() to clean up the outputs
+	initramfs.outputBuffer = new(bytes.Buffer)
+	initramfs.pgzWriter = pgzip.NewWriter(initramfs.outputBuffer)
+	initramfs.cpioWriter = cpio.NewWriter(initramfs.pgzWriter)
+
+	initramfs.initramfsOutputFile, err = os.OpenFile(initramfsPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, initramfsModeBits)
+
+	return
+}
+
+// OpenInitramfs makes an existing initramfs editable
+func OpenInitramfs(initramfsPath string) (initramfs InitramfsMount, err error) {
+	logger.Log.Debugf("Opening intramfs '%s'", initramfsPath)
+	inputFile, err := os.Open(initramfsPath)
+	if err != nil {
+		return
+	}
+	defer inputFile.Close()
+
+	gzReader, err := pgzip.NewReader(inputFile)
+	if err != nil {
+		return
+	}
+	defer gzReader.Close()
+	cpioReader := cpio.NewReader(gzReader)
+	//cpio.Reader has no Close() function to defer
+
+	// Caller must use InitramfsMount.Close() to clean up the outputs
+	initramfs.outputBuffer = new(bytes.Buffer)
+	initramfs.pgzWriter = pgzip.NewWriter(initramfs.outputBuffer)
+	initramfs.cpioWriter = cpio.NewWriter(initramfs.pgzWriter)
+
+	var bytesIO int64
+	var nextFileHeader *cpio.Header
+	for {
+		var linkPayload []byte
+
+		nextFileHeader, err = cpioReader.Next()
+		if err == io.EOF {
+			err = nil
+			break
+		}
+		if err != nil {
+			return
+		}
+
+		// For a given symlink, generate a payload that contains the read value of the link.
+		// The payload should be written after the header.
+		// e.g. A link from (/bin -> /usr/bin) would have a payload of "/usr/bin").
+		// cpio.ModeSymlink is two bits (one of which is resused), need to check for actual
+		// equality rather than non-zero after masking
+		isLink := (nextFileHeader.Mode&cpio.ModeType == cpio.ModeSymlink)
```

Comment:
>nextFileHeader.Mode&cpio.ModeType [](start = 13, length = 33)

Nit: this looks rahter odd and made me wonder if there's some Go vodoo at work here. Could you add spaces around "&" and maybe put the whole "X & Y"part in brackets?

Reply:
This is partly Linux Mode bit nonsense unfortunately. These aren't single bits, they are multi bit tags. While your way would work for symlinks specifically, it won't work generally. (Obnoxiously golang's version of the mode bits ARE single bit iotas... unfortunately the cpio library uses the linux ones raw from the header) 

---
In reply to: [550328663](https://github.com/microsoft/CBL-Mariner/pull/481#discussion_r550328663) [](ancestors = 550328663)

Reply:
I was able to add parentheses such that the linter accepts something a bit more human readable though. I think it tries to prioritize grouping to show order of operations more clearly so there is less ambiguity.

---
In reply to: [550696201](https://github.com/microsoft/CBL-Mariner/pull/481#discussion_r550696201) [](ancestors = 550696201,550328663)